### DISCUSSION TITLE
All error messages need to indicate which line of code in GDScript caused them to occur

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#8817] [Scripting]
One constant irritation working with Godot is the large number of error messages that indicate something going wrong in the C++ code and providing no indication which line in GDScript led to this happening.  

I'm getting lots of error messages like
```
./core/templates/cowdata.h:167 - Index p_index = -1 is out of bounds (size() = 0).
```
and

```
core/math/basis.cpp:47 - Condition "det == 0" is true.
core/math/transform_2d.cpp:51 - Condition "det == 0" is true.
```

There is no other indication of what is causing these errors.  My code base is pretty big now so I have little idea of where to look to start tracking these things down.  

It would be really nice if the error reporting mechanism could indicate which GDScript line was the entry point which leads to these errors.

**Comment by Mickeon** [comment]
Sometimes this is just not possible because the errors are happening internally during the lifetime of the Object and not as a direct result of the developer's code. For the times where they can be caught (such as `get_node()` not finding any nodes at a path), the source of the error is shown. So I am not sure where you're coming from, exactly. These feel like bugs.

**Comment by AThousandShips** [comment]
Are you actually sure these are caused by script code? If they aren't caused by a call directly from scripting, like happening inside the engine instead, then they don't have a script line, so the first step here is to confirm that

Generally errors caused by calls from gdscript are logged correctly, this would be easier to debug with:
* https://github.com/godotengine/godot-proposals/issues/963

**Comment by FabriceCastel** [comment]
I run into this problem _constantly_ with my gdscripts, I can't tell you how many times I ended up with a largely unhelpful `The axis Vector3 must be normalized` error coming out of the (moderately complex) animation scripts I'm working on. I would hope for this to point me back to the line in my scripts that set the problem vector instead of leaving me guessing.

I can't recall the other errors I've seen off the top of my head but if I run into them again I'll make a note of it. At the end of the day, I as a dev don't care for the particulars of whether my scripts caused the error directly or indirectly. I expect the error to point at the line in my script that caused it either way. I understand it's non-trivial, but it's currently my biggest gripe with Godot and would be the main reason I'd end up switching back to Unity if I ever did.

**Reply by AThousandShips** [reply]
> At the end of the day, I as a dev don't care for the particulars of whether my scripts caused the error directly or indirectly.

Well that's not really a justified complaint, a lot of the time there *isn't* a point in GDScript where the error "happened", unless it happens inside a call from GDScript there *isn't* a line that does it, for example, what line should be pointed to if you try to render an invalid mesh that's loaded from a scene file? There's no line that causes that, directly or indirectly, if for example it's the starting scene

If a broken mesh is rendered and an error happens, should it point to the place where the mesh was loaded? Where it was assigned to a `MeshInstance3D`? Where that instance was added to the tree?

Or if the error is because some model has an invalid transform, where should it point to? Where should an error like "network connection failed" point to?

In your case, where should it point to? The line where you select the animation that has the error? The line where it's loaded? Where is the line that "causes" this error? And how does it help you to know that?

*Most* of the cases that don't have a line are in fact these cases, because GDScript handles detecting this a lot of the time, there are some areas where it fails for various reasons, but those aren't necessarily trivial to fix either

But in the end this is not something that is broken, it's simply not something that *has* a solution, so please don't blow things out of proportion or present an impossible situation as something not being good enough 🙂

However there *are* areas where error messages could be more clear, *and* where stack traces would help a lot, like showing where an error happened within C++, like `Vector3::foo > Animation::bar > AnimationPlayer::baz`, and that's AFAIK what the proposal I linked above is about, but that also only really gives information if you know and understand the C++ code, so it's not necessarily any easier to identify

**Reply by Calinou** [reply]
The error messages could at least mention the invalid value you passed though, which would make troubleshooting a lot easier. I've started working on this a while ago for select error messages but it's a long and tedious process to catch all scenarios where values should be logged.

An ideal example is Image's `create_from_data()` error message, which I improved in https://github.com/godotengine/godot/pull/60126.

**Reply by AThousandShips** [reply]
Exactly 🙂

**Reply by Calinou** [reply]
Pull request opened: https://github.com/godotengine/godot/pull/87300

**Reply by blackears** [reply]
It could also be helpful if the path to the node which is causing the error is printed, as well as any script attached to the object in case it is a Resource or non-Node causing the error.  Just to give the person debugging it a bit more information.  At the moment I have Godot telling me about arrays being indexed at invalid locations and I have no clue where to start looking.

**Reply by AThousandShips** [reply]
When possible absolutely
